### PR TITLE
Enable row-group task partitioning in pyarrow-based read_parquet

### PR DESCRIPTION
**Pull request by rjzamora**
This PR adds the `split_row_groups` kwarg to `read_parquet`, which is used within the `"pyarrow"` engine to specify if output dataframe partitions should correspond to row groups (rather than distinct files).  In its current `read_metadata` definition, the `"pyarrow"` engine uses `pq.ParquetDataset.pieces`  to ultimately decide the granularity of `read_partition` tasks.  When the dataset is not *partitioned on columns*, these `pieces` each correspond to a file (note that, in contrast, the `"fastparquet"` engine uses row-groups, rather than files).

As discussed in [cudf#3142](https://github.com/rapidsai/cudf/issues/3142), in `dask_cudf` file-based task partitioning is problematic when the uncompressed file size is larger than GPU memory. 

**Notes**:
- `split_row_groups` is set to `True` by default (for consistency with fastparquet, and because this seems like a reasonable default)
- Row-group based task partitioning is disabled when `gather_statistics=False`, or when the dataset is already partitioned on 1+ columns (in the partitioned dataset case, the `pieces` are already smaller than the row-group).
- The logic added here can be dramatically simplified once the `split_row_groups` argument is implemented in pyarrow (for `pq.ParquetDataset`).  Although this argument [is already documented](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html?highlight=split_row_groups), using it results in `NotImplementedError`.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by quasiben** [comment]
@mrocklin do you have time to review this PR ?

**Comment by rjzamora** [comment]
Any thoughts on this @martindurant ?

File: dask/dataframe/io/parquet/arrow.py [hunk 1]
```diff
@@ -432,6 +475,7 @@ def write_partition(
                     metadata_collector=md_list,
                     **kwargs,
                 )
+            md_list[0].set_file_path(filename)
```
Comment (TomAugspurger) [inline dask/dataframe/io/parquet/arrow.py]: Do we know that `md_list` has at least one element at this point?

Comment (rjzamora) [reply dask/dataframe/io/parquet/arrow.py]: Good question - I added a check just in case something goes wrong.

**Comment by rjzamora** [comment]
> @rjzamora is https://issues.apache.org/jira/browse/ARROW-2801 the relevant JIRA for implementing split_row_groups in pyarrow? Could you comment on the sections of code that can be removed once it's implemented?

Yes, ARROW-2801 is the one.  I added a few brief "TODO" comments in the code to highlight the small changes that are needed once that issue is addressed in pyarrow.  In summary: Once `pq.ParquetDataset` can handle the `split_row_groups` kwarg, it can be directly passed within `_determine_dataset_parts` (and all new logic introduced in this PR can be removed, because `dataset.pieces` will already correspond to row-groups).

**Comment by martindurant** [comment]
So many special cases and flags :|  I hope to see this code disappear and be subsumed by pyarrow!

As far as it goes, I think the test alone proves it works well, so I am happy. I didn't gather, but for the dataset that is explicitly split, fastparquet already did the right thing here?

**Comment by rjzamora** [comment]
> As far as it goes, I think the test alone proves it works well, so I am happy. I didn't gather, but for the dataset that is explicitly split, fastparquet already did the right thing here?

Yes - It seems that fastparquet already partitions tasks by row-group.  "Fixing" this on the pyarrow side actually resolved many `pyarrow`-`fastparquet` compatibility issues in the tests.

**Pull request merged**